service/connection/ClientState.kt

```diff
@@ -141,14 +142,12 @@ class ClientState constructor(
     }
 
     private fun mayWarnPasswordExpiry() {
-        if (_client?.isCluster == true) {
-            val passwordExpiryDaysOptional: Optional<Long>? = _client?.asCluster()?.user()?.passwordExpiryDays()
-            if (passwordExpiryDaysOptional?.isPresent == true) {
-                val passwordExpiryDays = passwordExpiryDaysOptional.get()
-                if (passwordExpiryDays <= PASSWORD_EXPIRY_WARN_DAYS) {
-                    notificationSrv.userWarning(LOGGER, CREDENTIALS_EXPIRE_SOON, passwordExpiryDays)
-                }
-            }
+        if (_client?.isCluster == false) return
+        val passwordExpiryDurationOptional: Optional<Duration>? = _client?.asCluster()?.user()?.passwordExpirySeconds()?.map { Duration.ofSeconds(it) }
+        if (passwordExpiryDurationOptional?.isPresent == false) return
+        val passwordExpiryDuration = passwordExpiryDurationOptional?.get()
+        if (passwordExpiryDuration?.minus(PASSWORD_EXPIRY_WARN_DURATION)?.isNegative == true) {
+            notificationSrv.userWarning(LOGGER, CREDENTIALS_EXPIRE_SOON_HOURS, passwordExpiryDuration.toHours() + 1)
```

Comment:
There's a subtle mistake in your logic here which is causing us to need a lot more `?.` null-safe operators than we should.

`passwordExpiryDurationOptional?.isPresent == false` is only true when `passwordExpiryDurationOptional` is non-null and has a value of `false`. This means that we can hit L148 even when `passwordExpiryDurationOptional` is `null`.

To fix this, change `== false` to `!= true`.

Reply:
Very good catch, agreed. This is much easier to read.